@shopify/polaris Use rollup instead of loom for build process

alex-page opened the pull request.

alex-page committed:
Loom clean up

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Clean up typescript in workspaces

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Merge branch 'rm-ts-workspaces' into loom-to-rollup

alex-page committed:
migrate build

alex-page committed:
Move rollup to root package.json

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Merge branch 'mv-rollup-root' into loom-to-rollup

alex-page committed:
Delete loom.config.ts

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
rm loom

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Loom clean up

Co-Authored-By: Ben Scott <227292+BPScott@users.noreply.github.com>

aaronccasanova commented on polaris-react/config/rollup/plugin-styles.js:
```diff
@@ -119,14 +119,15 @@ module.exports.styles = function styles({
     }
 
     // Regular css file
+
```
```suggestion
```

aaronccasanova commented on polaris-react/rollup.config.js:
```diff
@@ -0,0 +1,96 @@
+      json({
+        compact: true,
+      }),
+
```
```suggestion
```

aaronccasanova commented on polaris-react/rollup.config.js:
```diff
@@ -0,0 +1,96 @@
+      },
+      {
+        format: 'esm',
+        dir: 'build/esm',
```
**suggestion (non-blocking):** Should we derive output `dir`s from the `package.json` like [we did in polaris-tokens](https://github.com/Shopify/polaris/blob/66b43a6892cdac31d6884c4dfd431c4588705d75/polaris-tokens/rollup.config.js#L20-L26)?

BPScott replied:
That's a nice idea! I think this being an absolute path would also mean that we can revert the `generateOptions` change in plugin-styles.

aaronccasanova commented on polaris-react/rollup.config.js:
```diff
@@ -0,0 +1,96 @@
+    output: [
+      {
+        format: 'esm',
+        dir: 'build/esnext',
```
Same as above

aaronccasanova commented on polaris-react/rollup.config.js:
```diff
@@ -0,0 +1,96 @@
+const {babel} = require('@rollup/plugin-babel');
```
Would be nice to [add `JSDoc`s](https://github.com/Shopify/polaris/blob/66b43a6892cdac31d6884c4dfd431c4588705d75/polaris-tokens/rollup.config.js#L11-L13) in this file to get some type hints throughout the config. Can be done in a follow up task.

BPScott replied:
💯 Love this. In an "improvements I'd like to see in all rollup config files" vein - Currently rollup.config.js files are transpiled through babel, and this leads to weirdness when importing commonjs files because how babel handles that vs how node handles that is not consistent (tis a hecking minefield). We can avoid all that by having `rollup.config.mjs` files that are pure esm, and won't get transpiled.

See https://rollupjs.org/guide/en/#configuration-files // https://rollupjs.org/guide/en/#using-untranspiled-config-files

aaronccasanova commented on package.json:
```diff
@@ -56,6 +56,14 @@
+    "@rollup/pluginutils": "^4.1.0",
     "stylelint": "^14.1.0",
     "turbo": "^1.2.8",
     "typescript": ">=3.0.0"
```
Do we need to adjust this `typescript` version range? Polaris was previously on `~4.3.5`. I realize that falls within `>=3`, but I'm not sure what version will be used now that `polaris-react`'s `typescript` dependency was removed.

BPScott replied:
Yep we do. `yarn why typescript` shows multiple versions of typescript installed. We should remove mentions of typescript on a package level and just have one definition at the root. This could be done outside of this PR, possibly in #5962, that pulls other packages up to the root

alex-page replied:
Yes. I agree a good follow up.

aaronccasanova reviewed (commented):
Nice work!! I'll pull down this branch in a bit to validate the upgrade 👍

alex-page committed:
Merge branch 'main' into mv-rollup-root

alex-page committed:
Update .eslintrc.js

alex-page committed:
changeset

BPScott committed:
Generating extraneous-dep configs

alex-page committed:
mv rollup

alex-page committed:
Merge branch 'mv-rollup-root' into loom-to-rollup

alex-page committed:
Update .eslintrc.js

alex-page committed:
Add config type

alex-page committed:
Update polaris-react/config/rollup/plugin-styles.js

Co-authored-by: Aaron Casanova <32409546+aaronccasanova@users.noreply.github.com>

alex-page committed:
Update polaris-react/rollup.config.js

Co-authored-by: Aaron Casanova <32409546+aaronccasanova@users.noreply.github.com>

alex-page committed:
Merge branch 'main' into loom-to-rollup

alex-page committed:
Update rollup.config.js

alex-page committed:
Update plugin-styles.js

alex-page committed:
Add rollup-plugin-node-externals

BPScott commented on polaris-react/scripts/build.js:
```diff
@@ -7,8 +7,14 @@ const run = (cmd) => execSync(cmd, {stdio: 'inherit', cwd: root});
-run(`yarn run loom build`);
+run(`yarn run rollup -c`);
+
+run('yarn run downlevel-dts "./build/ts/latest" "./build/ts/3.4"');
```
Some merge goof,  downlevel-dts / copyfiles commands are already present

BPScott commented on polaris-icons/rollup.config.js:
```diff
@@ -203,6 +203,9 @@ function replaceFillAttributeSvgoPlugin() {
 
+/**
+ * @type {import('rollup').RollupOptions}
+ */
```
Nitpick: you can specify these docblocks on one line

```suggestion
/** @type {import('rollup').RollupOptions} */
```

BPScott commented on polaris-react/rollup.config.js:
```diff
@@ -0,0 +1,100 @@
+
+/**
+ * @type {import('rollup').RollupOptions}
+ */
```
Nitpick: you can specify these docblocks on one line

```suggestion
/** @type {import('rollup').RollupOptions} */
```

BPScott commented on package.json:
```diff
@@ -57,14 +65,7 @@
+    "rollup-plugin-node-externals": "^4.0.0",
     "stylelint": "^14.1.0",
     "turbo": "^1.2.8",
     "typescript": ">=3.0.0"
```
I reckon this should be a bit nearer to reality rather than "any TS version released in the last 4 years".

```suggestion
    "typescript": "^4.3.5"
```

Run a `yarn` after this

BPScott reviewed (commented):
Looking great. A few small bits inline.

We should also delete the `polaris-react/loom.config.ts` file.

Looks like the build is failing due to a messed up import in the rollup config file.

alex-page committed:
Update polaris-icons/rollup.config.js

Co-authored-by: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Update polaris-react/rollup.config.js

Co-authored-by: Ben Scott <227292+BPScott@users.noreply.github.com>

alex-page committed:
Merge branch 'main' into loom-to-rollup

alex-page committed:
Update build.js

BPScott committed:
Merge remote-tracking branch 'origin/main' into loom-to-rollup

BPScott committed:
Fix things

BPScott committed:
remove loom config file

BPScott committed:
use package.json scripts instead of standalone file

BPScott committed:
Pure mjs rollup config

This means rollup doesn't transpile the contents of this file, it is
executed using node's esm handling.

BPScott committed:
Merge branch 'main' into loom-to-rollup

BPScott committed:
Revert typescript changes

BPScott committed:
Add jest types

BPScott reviewed (approved):
🎉
